File: backend/app/schemas/storyboard.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, computed_field, field_validator

from app.providers.speech.base import speech_seconds as estimate_speech_seconds
from app.schemas.environment import SceneEnvironment


class StoryboardSceneOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order: int
    duration_seconds: int
    # Null when the length is still derived from the dialogue. The panel
    # needs this to tell "Auto happened to pick 8s" from "somebody chose 8s",
    # which are the same number and different states.
    duration_override: int | None = None
    voiceover: str
    # When the creator last rewrote that line, or null while it is still the
    # agent's. Takes generated before it were made from words this scene no
    # longer says, and the preview marks them so.
    dialogue_edited_at: datetime | None = None
    visual_prompt: str
    caption: str
    features_creator: bool
    # Always present, even on scenes stored before setups existed: the
    # service fills the default in rather than making every caller handle a
    # null shape.
    environment: SceneEnvironment
    # Which take the final video uses. Zero-based.
    selected_take: int = 0
    # False for a scene the creator has left out of this cut.
    included_in_video: bool = True
    # True once the creator has edited the visual description by hand, which
    # is what stops a change of setup from quietly rewriting their words.
    visual_is_custom: bool

    @computed_field  # type: ignore[prop-decorator]
    @property
    def speech_seconds(self) -> float:
        """
        How long this line takes to say, by the same estimate the length is
        derived from. Sent rather than recomputed in the client so that
        "does it fit?" is answered by one words-per-second figure, not two
        that can drift apart.
        """
        return round(estimate_speech_seconds(self.voiceover), 2)

    @field_validator("environment", mode="before")
    @classmethod
    def _default_environment(cls, value: object) -> object:
        # Null on scenes stored before setups existed. Read as the default
        # rather than propagating the history into every client.
        return value or {}


class StoryboardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    qa_passed: bool
    qa_issues: list[str]
    scenes: list[StoryboardSceneOut]


class SceneOnCameraIn(BaseModel):
    features_creator: bool


class SceneDurationIn(BaseModel):
    """
    How long this one clip runs, or null to go back to fitting the dialogue.

    Validated against the provider rather than here: which lengths exist
    depends on the video provider and on whether the creator is in frame,
    and Veo answers both differently.
    """

    duration_seconds: int | None = None


class SceneInclusionIn(BaseModel):
    """Whether this scene goes into the finished video. The scene itself is
    kept either way - see storyboard_service.set_scene_inclusion."""

    included_in_video: bool


class SceneEnvironmentIn(BaseModel):
    """
    A scene's setup, replaced whole.

    `rebuild_visual` is the answer to "changing the setup may update the
    visual description": true rebuilds it, false keeps the creator's wording.
    It only matters for a scene whose visual was edited by hand; anything
    else rebuilds, since there is nothing there to protect.
    """

    environment: SceneEnvironment
    rebuild_visual: bool = True
    # Set when the creator picked a preset chip rather than tuning one
    # control. The server rebuilds the whole setup from that preset's
    # defaults, so "what YouTube Studio means" is defined in exactly one
    # place instead of being copied into the client.
    reset_to_preset: bool = False


class SceneDialogueIn(BaseModel):
    """
    What this scene says, rewritten by the creator.

    Nothing else here: the clip length and the prompt both follow the words
    on the server, because getting either of them wrong is what produces
    dead air or a clip talking over its own ending.
    """

    voiceover: str


class SceneVisualIn(BaseModel):
    """The visual description, written by the creator. Marks the scene custom
    so nothing rebuilds over it later without being asked."""

    visual_prompt: str


class ProjectEnvironmentIn(BaseModel):
    """
    The project's default setup.

    `apply_to_all` rewrites the setup on every existing scene. Off by
    default: a creator changing the default for scenes they have not written
    yet must not silently lose the setups they tuned on the ones they have.
    """

    environment: SceneEnvironment
    apply_to_all: bool = False
    reset_to_preset: bool = False

File: frontend/src/lib/api/storyboard.ts
import { api, ApiError } from "@/lib/api/client";
import type { Storyboard } from "@/types/storyboard";
import type { SceneEnvironment } from "@/types/environment";

export async function getStoryboard(projectId: string): Promise<Storyboard | null> {
  try {
    return await api.get<Storyboard>(`/projects/${projectId}/storyboard`);
  } catch (err) {
    if (err instanceof ApiError && err.status === 404) return null;
    throw err;
  }
}

// Backend has a single /generate endpoint that also serves as regenerate.
export async function generateStoryboard(projectId: string): Promise<Storyboard> {
  return api.post<Storyboard>(`/projects/${projectId}/storyboard/generate`);
}

export async function setSceneOnCamera(
  projectId: string,
  sceneId: string,
  featuresCreator: boolean
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}`,
    { features_creator: featuresCreator }
  );
}

/**
 * Replaces one scene's filming setup.
 *
 * `resetToPreset` says the creator clicked a preset chip rather than nudging
 * one control, so the server rebuilds the rest from that preset's defaults —
 * which is why "what YouTube Studio means" is not duplicated here.
 *
 * `rebuildVisual` decides what happens to a visual description the creator
 * wrote themselves. False keeps their words.
 */
export async function setSceneEnvironment(
  projectId: string,
  sceneId: string,
  environment: SceneEnvironment,
  options: { resetToPreset?: boolean; rebuildVisual?: boolean } = {}
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}/environment`,
    {
      environment,
      reset_to_preset: options.resetToPreset ?? false,
      rebuild_visual: options.rebuildVisual ?? true,
    }
  );
}

/** The creator's own wording for the shot. Nothing rebuilds over it after. */
export async function setSceneVisual(
  projectId: string,
  sceneId: string,
  visualPrompt: string
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}/visual`,
    { visual_prompt: visualPrompt }
  );
}

/**
 * Leaves a scene out of the finished video, or puts it back. The scene and
 * any clip it already has are kept, so this is free and reversible.
 */
export function setSceneInclusion(
  projectId: string,
  sceneId: string,
  includedInVideo: boolean
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}/inclusion`,
    { included_in_video: includedInVideo }
  );
}


/**
 * How long this one clip runs. Null goes back to fitting the dialogue.
 *
 * Only 4, 6 and 8 second clips exist, and only 8 while the creator is on
 * camera. The server refuses anything else rather than letting Veo reject
 * it mid-run, by which point earlier scenes have already been billed.
 */
export function setSceneDuration(
  projectId: string,
  sceneId: string,
  durationSeconds: number | null
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}/duration`,
    { duration_seconds: durationSeconds }
  );
}


/**
 * Rewrites what one scene says.
 *
 * The clip length and the prompt both follow the new words on the server,
 * so this is the only thing sent. Saving does not regenerate anything: the
 * clips already made are kept, marked as older than the line, and it is the
 * creator who decides which of them to replace.
 */
export function setSceneDialogue(
  projectId: string,
  sceneId: string,
  voiceover: string
): Promise<Storyboard> {
  return api.patch<Storyboard>(
    `/projects/${projectId}/storyboard/scenes/${sceneId}/dialogue`,
    { voiceover }
  );
}
